### ISSUE TITLE
WGSL built-in similar to `gl_DrawID` (for Multi Draw Indirect)

### ISSUE THREAD

**Issue by inodentry** [gfx-rs/wgpu#6823]
**Is your feature request related to a problem? Please describe.**

When using Multi Draw Indirect, there is no way for the shader to know what draw it is a part of.

GLSL in OpenGL provides the `gl_DrawID` built-in for this purpose. WGSL does not have an equivalent.

(AFAIK, the WebGPU spec does not have MDI at all? I am not sure. That would explain the lack of such builtin in WGSL.)

**Describe the solution you'd like**

Naga could add support for a `draw_index` built-in, similar to `vertex_index` and `instance_index`. Even though this is not part of the upstream WGSL spec.

**Describe alternatives you've considered**
Depending on the use case, it may be possible to work around this by using the instance IDs instead.

**Comment by jannik4** [comment]
I have also needed this for sometime and just used a fork. Currently it is just one line: https://github.com/jannik4/wgpu/commit/dce431a52938fc3b0e53a39eff4b0492fecb67f3

(But of course I would have no problem if that would be supported here:))

**Comment by cwfitzgerald** [comment]
Yeah, vulkan isn't the problem lol - we could support this on vulkan fairly easily, but metal and dx12 are significantly more work.

**Comment by jannik4** [comment]
Sure, I just restricted my app to Vulkan and called it a day😄.

**Comment by magcius** [comment]
If we can support versions of D3D12 with [incrementing constant](https://microsoft.github.io/DirectX-Specs/d3d/IndirectDrawing.html#incrementing-constant), then it shouldn't be too difficult to implement this, assuming we save a variable in the root signature somewhere for it.

Metal will be more complicated since it will require a buffer; I don't think you can set immediate data from an indirect command buffer.

**Comment by cwfitzgerald** [comment]
Well currently we don't use Indirect Command Buffers at all on metal, so MDI is actually just an single indirect call in a loop in the backend. That does actually make it slightly easier to do, as we can just set the immediate data.

**Comment by cwfitzgerald** [comment]
I've thought for a while we should have our own concept of an indirect command buffer in the shader to deal with these differences.

**Comment by ethanthoma** [comment]
I also need this. I want to use multi draw indirect and need per draw position offset (a vec3i). If I use a vertex attribute, it would double my memory usage.

**Comment by magcius** [comment]
Could you use instancing, and an instance-rate data stream?

**Comment by ethanthoma** [comment]
I use an instance buffer to represent a face of a voxel cube (draws 6 vertexes). My instance buffer has all my chunks in a single buffer. I use an indirect buffer to partially render my chunks. I store my chunk buffer data in a separate buffer that I want to index via the draw index (each draw call draws one chunk).

**Comment by ethanthoma** [comment]
Im not super familiar with graphics APIs so I'm not sure what a instance-rate data stream is. Thanks for your help!

**Comment by magcius** [comment]
Why not combine all your chunks into one giant indirect draw call using a compute shader? I don't see the need for multi-draw here.

**Comment by ethanthoma** [comment]
I could try that, thank you

**Comment by JonahPlusPlus** [comment]
This would be nice to have (I'm doing basically the same thing as @ethanthoma; [I think we both saw this video](https://www.youtube.com/watch?v=40JzyaOYJeY) at 8:30).

Edit: I think I'll just go with individual draw calls with push constants for now. Trying to use `draw_index` might be premature optimization anyways.

**Comment by tadeohepperle** [comment]
This would be extremely useful to have. I think there are quite many people just using wgpu as a Vulkan wrapper.

**Comment by Inspirateur** [comment]
Oh my god discovering that this is missing after 2 years of bevy voxel developpement is devastating lol.
This feature is very important for voxel games where the optimal setup is to use multi draw indexed indirect with 1 draw per chunk x face group, put the chunk x face group info in SSBO, and use the "draw index" to offset the quads correctly in order to render the game 😭 

All the workarounds to this are costly, in a setting where we can't afford it :(

**Comment by inner-daemons** [comment]
It looks like naga already supports this in the SPIR-V writer and SPIR-V parser. Unfortunately it isn't supported in the WGSL parser or HLSL/MSL writers however. I'd be willing to help push this through soon.

Just saw your message in the voxel discord @Inspirateur, sorry to hear about this rough edge. Thanks for the details about why its important also.

**Comment by inner-daemons** [comment]
Ok so in #8883 I've gotten this working on vulkan. Multi draw isn't available on metal so thats not a concern. DX12 I'd like to get these working for but it'd take some work.

Thats because this is a root constant in DX12, so we'd have to create a different root signature depending on whether this builtin is used, when currently the root signature only depends on the pipeline layout to my knowledge.

I don't have access to a windows computer very often anymore so don't expect that to come soon. But its in the back of my mind.

**Comment by cwfitzgerald** [comment]
> Multi draw isn't available on metal so thats not a concern

MDI is (emulated with an loop of draws), but MDIC isn't - so we will still need draw_id there, but we can do it using the same mechanism as immediate.

**Comment by inner-daemons** [comment]
@cwfitzgerald Ok, I would rather get multi draw indirect and multi draw indirect count working on metal before thinking about this though.
